source/crud/crud.rst

```diff
@@ -1953,54 +1953,37 @@ in the collection prior to ``$merge`` being executed).
-or ``$merge``. As of ``featureCompatibilityVersion`` 4.4, secondaries can now
-execute such pipelines; however, since drivers do not track
-``featureCompatibilityVersion``, executing an aggregation pipeline with ``$out``
-or ``$merge`` on a secondary will require a 5.0+ server (i.e. wire version >=
-13).
-
-Drivers MUST guarantee the following:
-
-- If an explicit (i.e. per-operation) read preference is specified for an
-  aggregation with a write stage, drivers MUST attempt to use it. If that would
-  result in a pre-5.0, secondary server being selected, drivers MUST instead
-  select a server using a primary read preference.
-- If no explicit read preference is specified but a default read preference is
-  available to inherit (e.g. from the Collection), drivers MUST attempt to use
-  it. If that would result in a pre-5.0, secondary server being selected,
-  drivers MUST instead select a server using a primary read preference.
-
-Drivers SHOULD use a custom server selector to consider server/wire version when
-matching a read preference and, if a pre-5.0 secondary would be selected, fall
-back to selecting a primary. With this approach, only a single server selection
-attempt is needed.
-
-If it is not possible to augment the server selector and a pre-5.0 secondary is
-selected, drivers MAY invoke server selection a second time to obtain a primary.
-However, if the first attempt results in a server selection timeout, drivers
-MUST NOT make a second attempt and MUST propagate the original timeout error. If
-the topology type is Single, Sharded, or LoadBalanced, drivers MUST NOT make a
-second attempt (there is no benefit in doing so).
-
-Regardless of whether drivers employ a custom server selector or two-attempt
-approach, they MUST discern the *effective* read preference for the operation,
-which SHALL be used for specifying the
+or ``$merge`` and drivers never considered a read preference for the operation.
+As of ``featureCompatibilityVersion`` 4.4, secondaries can now execute pipelines
+with ``$out`` or ``$merge``. Since drivers do not track
+``featureCompatibilityVersion``, the decision to consider a read preference for
+such a pipeline will depend on the wire version(s) of the server(s) to which the
+driver is connected.
+
+If there are one or more available servers and one or more of those servers is
+pre-5.0 (i.e. wire version < 13), drivers MUST NOT use the available read
+preference and MUST instead select a server using a primary read preference.
+Otherwise, if there are either no available servers or all available servers are
+5.0+ (i.e. wire version >= 13), drivers MUST use the available read preference.
+
+Drivers SHOULD augment their
+`server selection algorithm <..../server-selection/server-selection.rst#server-selection-algorithm>`_
+such that this logic can be enforced within a single server selection attempt.
+
+If the driver is connected to a load balancer, it MUST consider the wire version
```

Comment:
I don't think this is feasible.  Given the current CMAP spec, there is no guarantee that the driver will have a connection the LB available at the time when server selection is performed.  I think instead we should say that a driver can assume that any server behind an LB supports read preference for $out/$merge.

Reply:
This may overlap with a question @comandeo asked in Slack, to which I replied:

> Assuming the LB backs a 5.0+ server certainly works for this ticket, where you just need to determine if the server is pre- or post-5.0; however, you'll inevitably run into problems down the line if you need to do wire version checks for future features (e.g. something requires 5.2+). At some point, you will need to start making decisions on wire version reported in the handshake response.

Drivers are going to need to address this at some point. I don't think "defy augury" is a suitable excuse to avoid thinking about this.

Reply:
> Drivers are going to need to address this at some point. I don't think "defy augury" is a suitable excuse to avoid thinking about this.

Perhaps, but we don't have to do it right now in this PR.

Reply:
> Doesn't that still create a problem if we ultimately discover the driver is connected to a single 4.4 primary?

I don't think it does.  Since there are no servers available, it doesn't really matter what read preference will be applied, because the list of suitable servers will be empty, and driver will enter the server selection loop.  The next time the driver tries to apply the read preference, there will be a server available and we start the process again.

Reply:
> Perhaps, but we don't have to do it right now in this PR.

Just to confirm, you're OK with just removing this paragraph? If no other specs are talking about wire version checks for load balancers, I think leaving this unsaid is more future-proof.

> it doesn't really matter what read preference will be applied, because the list of suitable servers will be empty, and driver will enter the server selection loop. The next time the driver tries to apply the read preference, there will be a server available and we start the process again.

Understood. My mind is still thinking about the try-once mode for single-threaded selection. I didn't consider that the pre-5.0 server check would happen during _each_ loop iteration.

Reply:
Leaving it unsaid is more future-proof, but driver implementations will still need to special case this, so I think we should give guidance.  Something like:

> If the driver is connected to a load balancer. it MUST treat it the same as a 5.0+ cluster

Something like that.  What do you think?

Reply:
Updated the text to add a LoadBalanced topology to the list of cases where we apply the read preference.